Use dav1d decoder instead of libaom decoder

youennf opened the pull request:
#### 111902f17da1086c9096863b16d9ca382bf64f0a
<pre>
Use dav1d decoder instead of libaom decoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=248523">https://bugs.webkit.org/show_bug.cgi?id=248523</a>
rdar://problem/102804765

Reviewed by Jean-Yves Avenard.

dav1d decoder is more efficient than libaom decoder so it is best to use dav1d.
dav1d is a librarly linked in WebCore so we create a dav1d webrtc decoder in WebCore and not libwebrtc.
We integrate this decoder at WebKit level.

We have to export more headers, including libyuv from libwebrtc to be used in WebCore.
We update xcconfig for that purpose.

We link dav1d conditionnaly on ENABLE_WEB_RTC if it is not already linked through libavif.

* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOS.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOSsim.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.mac.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libyuv.xcconfig:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/WebKit/WebKitDecoder.h:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/WebKit/WebKitDecoder.mm:
(webrtc::RemoteVideoDecoder::RemoteVideoDecoder):
(webrtc::RemoteVideoDecoderFactory::CreateVideoDecoder):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCDefaultVideoDecoderFactory.m:
(-[RTCDefaultVideoDecoderFactory supportedCodecs]):
(-[RTCDefaultVideoDecoderFactory createDecoder:]):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCDefaultVideoEncoderFactory.m:
(+[RTCDefaultVideoEncoderFactory supportedCodecsWithH265:vp9Profile0:vp9Profile2:av1:]):
(-[RTCDefaultVideoEncoderFactory createEncoder:]):
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/libavif/dav1d/dav1d.xcodeproj/project.pbxproj:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDav1dDecoder.cpp: Added.
(WebCore::ScopedDav1dData::~ScopedDav1dData):
(WebCore::ScopedDav1dData::Data):
(WebCore::ScopedDav1dPicture::~ScopedDav1dPicture):
(WebCore::ScopedDav1dPicture::Picture):
(WebCore::NullFreeCallback):
(WebCore::Dav1dDecoder::Dav1dDecoder):
(WebCore::Dav1dDecoder::~Dav1dDecoder):
(WebCore::Dav1dDecoder::Configure):
(WebCore::Dav1dDecoder::RegisterDecodeCompleteCallback):
(WebCore::Dav1dDecoder::Release):
(WebCore::Dav1dDecoder::GetDecoderInfo const):
(WebCore::Dav1dDecoder::ImplementationName const):
(WebCore::Dav1dDecoder::Decode):
(WebCore::createLibWebRTCDav1dDecoder):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDav1dDecoder.h: Added.
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::createVideoDecoder):
(WebKit::releaseVideoDecoder):
(WebKit::decodeVideoFrame):
(WebKit::registerDecodeCompleteCallback):

Canonical link: <a href="https://commits.webkit.org/257283@main">https://commits.webkit.org/257283@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c70421c58ad7bc857add8632f88dcfdc3fdc1ed2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98437 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31566 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107863 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168129 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8157 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85035 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90985 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104517 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104103 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/89755 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33185 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88018 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/21114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1586 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/22642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1508 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5023 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6432 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42058 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2884 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDav1dDecoder.cpp:
```diff
@@ -0,0 +1,222 @@
+    Dav1dSettings s;
+    dav1d_default_settings(&s);
+
+    s.n_threads = std::max(2, settings.number_of_cores());
```
This is all very confusing. We set multi-thread decoding , yet have a max_frame_delay of 1 ; so the entire decoding later heavily rely on this configuration.

If this was changed later, the entire code below would become unsafe, as dav1d could spawn another thread to perform the decoding on the compressed frame and queue it for later.

I think this deserves a comment so that it's not changed without understanding the consequences.

youennf replied:
Will add it. And we should indeed allow to use multithreading in a follow-up.

jyavenard reviewed (approved):
r=me with nit

youennf committed:
Use dav1d decoder instead of libaom decoder
https://bugs.webkit.org/show_bug.cgi?id=248523
rdar://problem/102804765

Reviewed by Jean-Yves Avenard.

dav1d decoder is more efficient than libaom decoder so it is best to use dav1d.
dav1d is a librarly linked in WebCore so we create a dav1d webrtc decoder in WebCore and not libwebrtc.
We integrate this decoder at WebKit level.

We have to export more headers, including libyuv from libwebrtc to be used in WebCore.
We update xcconfig for that purpose.

We link dav1d conditionnaly on ENABLE_WEB_RTC if it is not already linked through libavif.

* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOS.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.iOSsim.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.mac.exp:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.xcconfig:
* Source/ThirdParty/libwebrtc/Configurations/libyuv.xcconfig:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/WebKit/WebKitDecoder.h:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/WebKit/WebKitDecoder.mm:
(webrtc::RemoteVideoDecoder::RemoteVideoDecoder):
(webrtc::RemoteVideoDecoderFactory::CreateVideoDecoder):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCDefaultVideoDecoderFactory.m:
(-[RTCDefaultVideoDecoderFactory supportedCodecs]):
(-[RTCDefaultVideoDecoderFactory createDecoder:]):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCDefaultVideoEncoderFactory.m:
(+[RTCDefaultVideoEncoderFactory supportedCodecsWithH265:vp9Profile0:vp9Profile2:av1:]):
(-[RTCDefaultVideoEncoderFactory createEncoder:]):
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/libavif/dav1d/dav1d.xcodeproj/project.pbxproj:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDav1dDecoder.cpp: Added.
(WebCore::ScopedDav1dData::~ScopedDav1dData):
(WebCore::ScopedDav1dData::Data):
(WebCore::ScopedDav1dPicture::~ScopedDav1dPicture):
(WebCore::ScopedDav1dPicture::Picture):
(WebCore::NullFreeCallback):
(WebCore::Dav1dDecoder::Dav1dDecoder):
(WebCore::Dav1dDecoder::~Dav1dDecoder):
(WebCore::Dav1dDecoder::Configure):
(WebCore::Dav1dDecoder::RegisterDecodeCompleteCallback):
(WebCore::Dav1dDecoder::Release):
(WebCore::Dav1dDecoder::GetDecoderInfo const):
(WebCore::Dav1dDecoder::ImplementationName const):
(WebCore::Dav1dDecoder::Decode):
(WebCore::createLibWebRTCDav1dDecoder):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCDav1dDecoder.h: Added.
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.cpp:
(WebKit::createVideoDecoder):
(WebKit::releaseVideoDecoder):
(WebKit::decodeVideoFrame):
(WebKit::registerDecodeCompleteCallback):

Canonical link: https://commits.webkit.org/257283@main
